File: ruby/twelve-days/twelve_days.rb
# Twelve Days of Christmas exercise on exercism
class TwelveDays
  DAYS_GIFTS = [
    ['first', 'a Partridge in a Pear Tree'],
    ['second', 'two Turtle Doves'],
    ['third', 'three French Hens'],
    ['fourth', 'four Calling Birds'],
    ['fifth', 'five Gold Rings'],
    ['sixth', 'six Geese-a-Laying'],
    ['seventh', 'seven Swans-a-Swimming'],
    ['eighth', 'eight Maids-a-Milking'],
    ['ninth', 'nine Ladies Dancing'],
    ['tenth', 'ten Lords-a-Leaping'],
    ['eleventh', 'eleven Pipers Piping'],
    ['twelfth', 'twelve Drummers Drumming']
  ].freeze

  GIFTS = DAYS_GIFTS.map { |_, gift| gift }

  def self.song
    new.song
  end

  def song
    DAYS_GIFTS.map.with_index do |(day, _), index|
      verse(day, day_gifts(index))
    end.join("\n")
  end

  private

  def verse(day, gifts)
    "On the #{day} day of Christmas my true love gave to me: #{gifts}.\n"
  end

  def day_gifts(index)
    *gifts, last_gift = GIFTS[0..index].reverse
    return last_gift if index.zero?

    (gifts + ["and #{last_gift}"]).join(', ')
  end
end

File: haskell/twelve-days/src/TwelveDays.hs
module TwelveDays (recite) where

recite :: Int -> Int -> [String]
recite start stop = map sentence dates
  where dates = [start..stop]

sentence :: Int -> String
sentence d = "On the " ++ date ++ " day of Christmas my true love gave to me: " ++ gotten ++ "."
  where gotten = foldl1 (\acc p -> acc ++ ", " ++ p) (goodies d)
        date = day d

goodies :: Int -> [String]
goodies n 
  | n == 1 = [got n]
  | n == 2 = got n : ["and " ++ got (n - 1)]
  | otherwise = got n : goodies (n - 1)

day :: Int -> String
day d = case d of
          1   -> "first"
          2   -> "second"
          3   -> "third"
          4   -> "fourth"
          5   -> "fifth"
          6   -> "sixth"
          7   -> "seventh"
          8   -> "eighth"
          9   -> "ninth"
          10  -> "tenth"
          11  -> "eleventh"
          12  -> "twelfth"
          _   -> "other"

got :: Int -> String
got d = case d of
          1   -> "a Partridge in a Pear Tree"
          2   -> "two Turtle Doves"
          3   -> "three French Hens"
          4   -> "four Calling Birds"
          5   -> "five Gold Rings"
          6   -> "six Geese-a-Laying"
          7   -> "seven Swans-a-Swimming"
          8   -> "eight Maids-a-Milking"
          9   -> "nine Ladies Dancing"
          10  -> "ten Lords-a-Leaping"
          11  -> "eleven Pipers Piping"
          12  -> "twelve Drummers Drumming"
          _   -> "Nothing"

